src/model/operation/markeroperation.js

```diff
@@ -0,0 +1,152 @@
+/**
+ * @license Copyright (c) 2003-2017, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/**
+ * @module engine/model/operation/markeroperation
+ */
+
+import Operation from './operation';
+import Range from '../range';
+import CKEditorError from '@ckeditor/ckeditor5-utils/src/ckeditorerror';
+
+/**
+ * @extends module:engine/model/operation/operation~Operation
+ */
+export default class MarkerOperation extends Operation {
+	/**
+	 * @param {String} name Marker name.
+	 * @param {module:engine/model/range~Range} oldRange Marker range before the change.
+	 * @param {module:engine/model/range~Range} newRange Marker range after the change.
+	 * @param {Number} baseVersion {@link module:engine/model/document~Document#version} on which the operation can be applied.
+	 */
+	constructor( name, oldRange, newRange, baseVersion ) {
+		super( baseVersion );
+
+		/**
+		 * Marker name.
+		 *
+		 * @readonly
+		 * @member {String}
+		 */
+		this.name = name;
+
+		if ( ( oldRange && !oldRange.root.document ) || ( newRange && !newRange.root.document ) ) {
+			/**
+			 * MarkerOperation range must be inside a document.
+			 *
+			 * @error marker-operation-range-not-in-document
+			 */
+			throw new CKEditorError( 'marker-operation-range-not-in-document: MarkerOperation range must be inside a document.' );
+		} else if ( oldRange && newRange && oldRange.root.document != newRange.root.document ) {
+			/**
+			 * MarkerOperation ranges must be inside same document.
+			 *
+			 * @error marker-operation-ranges-in-different-documents
+			 */
+			throw new CKEditorError(
+				'marker-operation-ranges-in-different-documents: MarkerOperation ranges must be inside same document.'
+			);
+		}
+
+		/**
+		 * Marker range before the change.
+		 *
+		 * @readonly
+		 * @member {module:engine/model/range~Range}
+		 */
+		this.oldRange = oldRange ? Range.createFromRange( oldRange ) : null;
+
+		/**
+		 * Marker range after the change.
+		 *
+		 * @readonly
+		 * @member {module:engine/model/range~Range}
+		 */
+		this.newRange = newRange ? Range.createFromRange( newRange ) : null;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	get type() {
+		return 'marker';
+	}
+
+	/**
+	 * @inheritDoc
+	 * @returns {module:engine/model/operation/markeroperation~MarkerOperation}
+	 */
+	clone() {
+		return new MarkerOperation( this.name, this.oldRange, this.newRange, this.baseVersion );
+	}
+
+	/**
+	 * @inheritDoc
+	 * @returns {module:engine/model/operation/markeroperation~MarkerOperation}
+	 */
+	getReversed() {
+		return new MarkerOperation( this.name, this.newRange, this.oldRange, this.baseVersion + 1 );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	_execute() {
+		// If old range and new range are "same", operation should not do anything.
+		if ( this.oldRange === null && this.newRange === null ) {
+			return;
+		} else if ( this.oldRange !== null && this.newRange !== null && this.oldRange.isEqual( this.newRange ) ) {
+			return;
+		}
```

Comment:
As we found out, now if you add existing marker to the document the change event won't fire, what is wrong.